### PR TITLE
Create `MotionMintTokens` saga

### PR DESCRIPTION
**Pull request by chinins**
## Description

This PR adds actions and saga for mint tokens motion.

It adds colonyNetwork reputation changes. It also adds script to add proxy for the reputationOracleEndpoint. There is a need for a change in the colony-js but Alex said that we can't do a new release just for that, so he added proxy.

The steps to test:

- `npm i`
- `npm run provision`
- start `dev-heavy` environment
- create a colony
- mint tokens
- make a payment. You need to wait a bit for the reputation to get updated and also reload the web app. It caught me several times. This needs to be investigated more but it's out of the scope of this PR. The motion saga will not work without any reputation in the colony.
- install & enable voting extension
- **mint tokens  - that would trigger `mintTokensMotion`**

**New stuff** ✨

- add mint tokens motion actions
- add mint tokens motion saga
- add condition to mint tokens dialog to check if the voting extension is installed

**Deletions** ⚰️

- none

## TODO

- Prevent mint tokens motion saga to start if there is no reputation in the colony

Resolves DEV-251

image

### CODE REVIEW

**Comment by rdig** [comment]
I mean... we already have a mint tokens saga.

This will be more like `MotionMintTokens` saga

Also, the redux action name should be `COLONY_MOTION_MINT_TOKENS` _(The `VOTE` naming is wrong since we are going to have an **actual** vote transaction that we need to fire)_

**Comment by chinins** [comment]
@rdig , thank you for the comments - will update the naming

**Review by ArmandoGraterol — COMMENTED** [review]
The code seems ok to me, however, when I replicated the steps you wrote to test it I get that I don't have the root permission needed to mint tokens

**Review by rdig — CHANGES_REQUESTED** [review]
This is very good work.

Screenshot from 2021-03-24 20-47-52

On one hand it works perfectly _(with some exception, but are things that you didn't introduce here)_, on the other I found some code mistakes and improvements that I want you to do.

As I've mentioned above, while this is not something you introduced, I still noticed it here, so since you're at it please fix it.

@alicjakujawa introduced a check in `ManageFundsDialog` that requires permissions if the voting reputation extension is not installed. This is actually not correct, since it makes that if the extension is not installed, you cannot mint tokens anymore.

Screenshot from 2021-03-24 20-42-37

Please remove that check.

I'll re-review once everything is corrected.

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx [hunk 1]
```diff
@@ -94,27 +116,33 @@ const TokenMintDialog = ({
...
-            isVotingExtensionEnabled={isVotingExtensionEnabled}
-            back={prevStep && callStep ? () => callStep(prevStep) : undefined}
-            nativeToken={nativeToken}
-          />
-        </Dialog>
-      )}
+      {(formValues: FormikProps<FormValues>) => {
+        if (formValues.values.forceAction !== isForceOn) {
+          setIsForceOn(formValues.values.forceAction);
+        }
```
[the remark this answers is not in the document]
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx:133]: This for the force toggle button. If I don't keep it in state and change as we go, the actions that `submit` will trigger will not react to the the toggle and might be wrong.

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx [hunk 2]
```diff
@@ -61,8 +64,27 @@ const TokenMintDialog = ({
...
   prevStep,
 }: Props) => {
+  const [isForceOn, setIsForceOn] = useState(false);
   const history = useHistory();
 
+  const { data } = useColonyExtensionsQuery({
+    variables: { address: colonyAddress },
+  });
+
+  const hasVotingExtension = useHasVotingExtension(data);
```
Comment (alicjakujawa) [inline src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: This is not scalable. I created something similar - useEnabledExtensions hook. We should have hook which handle all cases, not for each extension one hook. So if my hook doesn't work for you (or has bug), just refactor it to make it more useful across the app

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (alicjakujawa) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: Im still not sure why we're doing it 🤔

File: src/modules/dashboard/sagas/actions/mintTokensMotion.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    } = yield takeFrom(
+      createRootMotion.channel,
+      ActionTypes.TRANSACTION_HASH_RECEIVED,
+    );
+    yield takeFrom(createRootMotion.channel, ActionTypes.TRANSACTION_SUCCEEDED);
+
+    if (annotationMessage) {
+      yield put(transactionPending(annotateMintTokens.id));
+
+      let ipfsHash = null;
```
Comment (alicjakujawa) [inline src/modules/dashboard/sagas/actions/mintTokensMotion.tsx:117]: why let in here?

**Review by alicjakujawa — CHANGES_REQUESTED** [review]

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 2 above)
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: Yeah, already saw and changed - there is already an updated version

File: src/modules/dashboard/sagas/actions/mintTokensMotion.tsx (hunk 3 above)
Comment (chinins) [reply src/modules/dashboard/sagas/actions/mintTokensMotion.tsx:117]: It's used like this in all other sagas, so I just implemented the same. But it's a good question indeed... myabe @rdig  can tell us if it's needed. And I can remove if not.

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: `isForce` is used to determine which action we trigger when hitting submit. This value we take from our form in `TokenMintForm`. How else do we get the value of force toggle from the form? Maybe there is another way that I don't know about?

[the remark this answers is not in the document]
Comment (alicjakujawa) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: I mean im not sure why we're adding it to the local state if its hold in formValues?

File: src/modules/dashboard/sagas/actions/mintTokensMotion.tsx (hunk 3 above)
Comment (rdig) [reply src/modules/dashboard/sagas/actions/mintTokensMotion.tsx:117]: I mean, the explanation for this is right on the next line :)

https://github.com/JoinColony/colonyDapp/blob/eed7ed60e0b52c7f5c693b117b6af5eaf0aafd6c/src/modules/dashboard/sagas/actions/mintTokensMotion.tsx#L118-L123

The thing is that of the IPFS upload call fails it never sets a value to `ipfsHash` and when you try to call the contract method with the `ipfsHash` param being `undefined`, it will revert with an error.

But if it's `null` the call will succeed (obviously without an ipfs hash being set)

This is just basic fallback method to ensure that the call will always work, even if IPFS fails.

We'll improve on this in the future, but for the time being, this is good enough,

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: How am I going to use `formValues` outside the function call where it's provided? What am I missing?

File: src/modules/dashboard/sagas/actions/mintTokensMotion.tsx (hunk 3 above)
Comment (chinins) [reply src/modules/dashboard/sagas/actions/mintTokensMotion.tsx:117]: I suspected something like this but was not sure. Thank you for the explanation!

**Comment by alicjakujawa** [comment]
Screenshot 2021-03-25 at 16 19 41
@chinins so this error is because i run mint tokens without any reputation? If so i think we need to somehow handle this cases, to not throw error on user face :D Can we disable button with description where there is 0 reputation? 🤔

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rdig) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: I think Alicja is thinking about using the actual form values available inside the `submit`/`error`/`success` calls.

So something like this:
```js
      submit={formValues => getFormAction('SUBMIT', formValues.forceAction)}
```

And then isnde the `getFormAction`:
```js
  const getFormAction = useCallback(
    (actionType: 'SUBMIT' | 'ERROR' | 'SUCCESS', forceState) => {
    ...
  });
```

**Comment by rdig** [comment]
> @chinins so this error is because i run mint tokens without any reputation? If so i think we need to somehow handle this cases, to not throw error on user face :D Can we disable button with description where there is 0 reputation? thinking

@alicjakujawa We have a issue to solve this: https://linear.app/colony/issue/DEV-268/prevent-triggering-a-motion-when-colony-reputation-is-0

It should stay like this for now, as we don't have a proper solution for this.

**Comment by chinins** [comment]
> Screenshot 2021-03-25 at 16 19 41
> 
> @chinins so this error is because i run mint tokens without any reputation? If so i think we need to somehow handle this cases, to not throw error on user face :D Can we disable button with description where there is 0 reputation? 🤔

@alicjakujawa, in my PR description in steps to test I specifically write about this situation. There is a separate issue that tracks it and will solve this problem as we are not going to do it in this PR as we need some more inputs from the product.
https://linear.app/colony/issue/DEV-268/prevent-triggering-a-motion-when-colony-reputation-is-0

This morning I posted in our frontend group chat an update about it:
`Just to give you an update about the DEV-268 Prevent triggering a motion when colony reputation is 0, this is what Karol sent me:
"- on our side, we will add additional warning to Permissions UAC modal in case the user would be the last person who wants to remove permissions - we will have more clear description when task is ready; it is in my backlog and most likely, I will tackle it in next few days; 
- I am still skeptical about blocking the possibility to create a motion in case you have provided, especially that there is risk of having funds locked; however we might for the current time being disable it and just flag the possible issue"
So to sum up, it's on standby for the moment. Karol will provide us with warning for Permissions modal & the UI for flagging the issue when we prevent creating a motion.`

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: Ok, thank you! Now I see what you mean. Sure, I can refactor, I don't mind.

**Review by rdig — APPROVED** [review]
Good to go! Very nice work!

File: src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (chinins) [reply src/modules/dashboard/components/TokenMintDialog/TokenMintDialog.tsx]: Just to keep everybody in the loop - we can't do Raul's suggestion as it requires refactoring `ActionForm` component & we agreed that it's not worth it.

File: scripts/start_all.js [hunk 4]
```diff
@@ -129,7 +150,7 @@ addProcess('graph-node', async () => {
   await new Promise(resolve => {
     console.log(); // New line
     console.log('Cleaning up the old graph-node docker data folder. For this we need', chalk.bold.red('ROOT'), 'permissions');
-    sudo.exec(`rm -Rf ${path.resolve(__dirname, '..', 'src/lib/graph-node/docker/data')}`, {},
+    sudo.exec(`rm -Rf ${path.resolve(__dirname, '..', 'src/lib/graph-node/docker/data')}`, {name: 'GraphNodeCleanup'},
```
[the remark this answers is not in the document]
Comment (area) [reply scripts/start_all.js:153]: Mostly makes it work for me on my new machine. `process.title` has something wrong with it, and so I hit [this error](https://github.com/jorangreef/sudo-prompt/blob/master/index.js#L62). This was just the quickest (selfish) solution :stuck_out_tongue:

File: src/lib/reputationMonitor/index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+const colonyNetwork = new ethers.Contract(
+  colonyNetworkAddress,
+  networkAbi,
+  provider,
+);
+
+async function forwardTime(seconds) {
+  await provider.send('evm_increaseTime', [seconds]);
+  await provider.send('evm_mine');
+}
```
[the remark this answers is not in the document]
Comment (area) [reply src/lib/reputationMonitor/index.js:23]: Your own `curl` commands should still cause a reputation cycle to occur, which will cause reputation to decay (though probably not noticeably).
